### ISSUE TITLE
System commands as message filters

### ISSUE THREAD

**Issue by foxcpp** [foxcpp/maddy#96]
While it should be noted that this mechanism does not scale well to large amounts of messages, it might be useful in simple configurations where this scalability is not required.

```
filter command {
  cmd spamc
}
```
The specified command will have the message header and body copied to its stdin. New header and body will be read from stdout.
If the commands exit status is not zero - the message will be rejected.

"cmd" directive may specify additional arguments to the command. The command is executed directly, not using a system shell to avoid unexpected arguments expansion. We might add placeholders for passing envelope information in headers, like `cmd /bin/check_sender $from` but they should be used with extreme care to avoid the possibility of RCE.

Optional directives that might be useful:
* `stdout_header_only`
  Command only prints new header, not the entire body. Useful to avoid copying the entire body around, rebuffering it and etcetera.
* `stdin_header_only`
  Copy only message header to stdin. Useful if the command doesn't care about the message body.
* `ignore_fail`
  Ignore non-zero exit code and continue processing. Similar to check_source_* filters family.
* `error_response <code> [message]`
  SMTP status code and message to return if command fails. Without it generic response is returned, probably `451 Message rejected by filter`

**Comment by emersion** [comment]
Looks sane to me

**Comment by foxcpp** [comment]
And yeah, if command stdout is empty, it will not affect the message.

```
filter command {
  cmd grep -Piv "you have \d profile view waiting for you"
}
```

**Comment by foxcpp** [comment]
>And yeah, if the command's stdout is empty, it will not affect the message.

Or, well, probably it is better to have an explicit option like `stdout_ignore`.

**Comment by foxcpp** [comment]
Revisiting design for new interfaces (#108) and "actions" from #119.
```
check {
  command {
    sender_cmd ./check_sender $address
    sender_fail_action quarantine # optional?
    sender_error 500 5.0.0 "Go away" # optional

    rcpt_cmd ./check_rcpt $address
    rcpt_fail_action quarantine # optional?
    rcpt_error 500 5.0.0 "Go away" # optional

    body_cmd ./check_body
    body_header_only yes # defaults to 'no'
    body_fail_action quarantine # optional?
    body_error 500 5.0.0 "Go away" # optional
  }
}

Placeholders:
- {address} - sender or recipient address, depending on the command (sender_cmd vs rcpt_cmd)
- {sender} - sender address
- {client_ehlo} - hostname provided by the client in the EHLO command
- {client_ip}
- {msg_id} - MsgMeta.MessageID
```

**Comment by foxcpp** [comment]
sign_dkim uses `{name}` for placeholders in config. Updating proposals in this issue to use the same format for consistency.

Adding to 0.1 milestone, system commands provide an easy way to integrate external filtering software with maddy and given the intention to limit amount of built-in filters, this becomes an important part of functionality.

**Comment by foxcpp** [comment]
Closed in d80c04d.
